waves/src/App.tsx

```diff
@@ -229,6 +223,18 @@ function App() {
         state.wallet.status.loaded ? 5000 : null,
     );
 
+    state.rate = useSSE("rate", {
+        ask: 19133.74,
+        bid: 19133.74,
+    });
+
+    useEffect(() => {
+        dispatch({
+            type: "UpdateBetaAmount",
+            value: state.rate,
+        });
+    }, [state.rate]);
```

Comment:
Non-urgent feedback, more of a long-term thing to consider:

It is a bit odd that you have to trigger another effect here based on the rate change because we already have a reducer-like construct within `useSSE`.

I think this stems from your reducer calculating both amounts. That is in the end more state than you should keep in your reducer (i.e. it is not normalized because one field depends on the other).

I feel it should be possible to model this in a more functional way where the only thing you store are the input values (and probably which field the user touched last) and you dynamically compute the other inputs field based on the rate. By dynamically computing I mean outside of the reducer, simply as part of the render function.

That would allow you to remove the rate from the reducer and you can just use the variable returned from `useSSE` which in turn makes this `useEffect` go away!

A component (in particular the render function but you only have function components so here it is equivalent) is always a pure function of data (props + state) to a view. Any kind of side-effect is either outside or modeled through hooks.
It makes things a lot easier to understand if you use side-effects only for the things that have to be side-effects. For example, from a domain-perspective, one of the input's values is always a pure function of the rate + the other input. As such, computing its value could be done as part of the render function.

Reply:
I'm not sure I follow. 

The beta amount is always `alpha`*`rate` and cannot be edited. This was changed when you mentioned it the first time :) 

So to the rate, `useSSE` returns the rate. How should I trigger an update of the `beta`-amount if not with the reducer?

Reply:
Ah, I think I have an idea...

Reply:
checkout latest commit: [`9d36876` (#46)](https://github.com/comit-network/droplet/pull/46/commits/9d368769592eeb7bc9403e2c466c3f667fbd9091). I'm not sure how to get rid of `state.rate = useSSE` as I store the rate in the state.

Reply:
The last patch looks almost like what I meant :)

Why does the rate have to be in the state though?

Just make it a local variable of your component!

Reply:
If the rate has to be in the state because you need it in the reducer for handling other actions, you should introduce the action again that dispatches the rate.